app/models/report.rb

```diff
@@ -3,15 +3,33 @@
 class Report < ApplicationRecord
   belongs_to :user
   has_many :comments, as: :commentable, dependent: :destroy
+  has_many :active_relationships, class_name: 'Relationship', foreign_key: 'mentioning_id', inverse_of: :mentioning, dependent: :destroy
+  has_many :passive_relationships, class_name: 'Relationship', foreign_key: 'mentioned_id', inverse_of: :mentioned, dependent: :destroy
+  has_many :mentioning_reports, through: :active_relationships, source: :mentioned
+  has_many :mentioned_reports, through: :passive_relationships, source: :mentioning
 
   validates :title, presence: true
   validates :content, presence: true
 
+  after_create :save_mentions
+  after_update :delete_mentions, :save_mentions
+
   def editable?(target_user)
     user == target_user
   end
 
   def created_on
     created_at.to_date
   end
+
+  private
+
+  def save_mentions
+    mentioned_report_ids = content.scan(%r{http://localhost:3000/reports/(\d+)}).flatten.map(&:to_i)
```

Comment:
非常に細かいのですが、`(has_manyのモデル名)_ids`という名前にするとRailsが自動的に作成するメソッドと名前が被ります。
ためしに以下のコードを実行してみてください。

```suggestion
    puts mentioned_report_ids # 変数を宣言する前でも呼び出せる = Railsが自動的に作成するメソッドが呼ばれる
    mentioned_report_ids = content.scan(%r{http://localhost:3000/reports/(\d+)}).flatten.map(&:to_i)
    puts mentioned_report_ids # ここからは変数の内容に切り替わる
```

参考 https://qiita.com/jnchito/items/fb438aa0786bb3cbf919

誤解を招きやすくなるので別の名前を考えてみてください。
（寿命が短い変数なので、もっと短くてもよいです）

Reply:
idsに修正しました。